# Segmenting and Clustering Neighborhoods in Toronto 3

First we import all the libraries we probably are going to use

In [122]:

import numpy as np 
import pandas as pd 
from pandas.io.json import json_normalize
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium # map rendering library
from bs4 import BeautifulSoup

print('Libraries imported.')

Libraries imported.


In [148]:
{
    "tags": [
        "hide-cell",
    ]
}
wikipedia_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

soup = BeautifulSoup(wikipedia_url, 'html5lib')

table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

import geocoder

df_coords = pd.read_csv('Geospatial_Coordinates.csv')
df_coords.rename({'Postal Code': 'PostalCode'}, axis=1, inplace=True)
merged = pd.merge(df, df_coords)

# Start the analysis

In [149]:
{
    "tags": [
        "hide-cell",
    ]
}

CLIENT_ID = 'ZKGM0MFL1DUXQSPJUSZ5ODBKK5PBSOW5O0QSFHXE2BFDS0MW' # your Foursquare ID
CLIENT_SECRET = 'PKGKDW03TL1FMDXLKEWY4PCLSA1LBCDOOKP433XGPYZXHHE2' # your Foursquare Secret
ACCESS_TOKEN = '2KU2YL54Q51T2QWYXFYLBWPXVOA3GP2KSJXDBDP2JLJF522H' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30

#### I first find the geograpical coordinates of Toronto

In [125]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Then plot a map with all the neighborhoods identified

In [126]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(merged['Latitude'], merged['Longitude'], merged['Borough'], merged['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Created a funtion that returns all the venues in a neighborhood

In [127]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Then used the past funtion to create a dataframe with the rows(Neighborhood	Neighborhood Latitude	Neighborhood Longitude	Venue	Venue Latitude	Venue Longitude	Venue Category)

In [128]:
toronto_venues = getNearbyVenues(names=merged['Neighborhood'],
                                   latitudes=merged['Latitude'],
                                   longitudes=merged['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [129]:
print(toronto_venues.shape)
toronto_venues.head()

(1273, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0                     KFC       43.754387       -79.333021   
1         Brookbanks Park       43.751976       -79.332140   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0   Fast Food Restaurant  
1                   Park  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

#### We found 1273 venues in Toronto

#### Then count number of venues in each Neighborhood

In [130]:
venues_count = toronto_venues.groupby('Neighborhood').count()
venues_count

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Wilson Heights, Downsview North                        15   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            30   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           22   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               7   
Central Bay Street                                                     30   
Christie                                                               15   
Church and Wellesley                                                   30   
Clairville, Humberwood, Woodbine Downs, West Hu...                      4   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             28   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills North                                                         5   
Don Mills South                                                        16   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview Central                                                       3   
Downsview East                                                          4   
Downsview Northwest                                                     6   
Downsview West                                                          4   
Dufferin, Dovercourt Village                                           14   
Enclave of L4W                                                         14   
Enclave of M4L                                                         19   
Enclave of M5E                                                         30   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      7   
Fairview, Henry Farm, Oriole                                           30   
First Canadian Place, Underground city                                 30   
Forest Hill North & West                                                4   
Garden District, Ryerson                                               30   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       9   
Harbourfront East, Union Station, Toronto Islands                      30   
High Park, The Junction South                                          18   
Hillcrest Village                                                       6   
Humber Summit                                                           2   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         19   
Kennedy Park, Ionview, East Birchmount Park                             7   
Kensington Market, Chinatown, Grange Park                              30   
Kingsview Village, St. Phillips, Martin Grove G...                      3   
Lawrence Manor, Lawrence Heights                

In [131]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 227 uniques categories.


#### Then find how many venues there is for each category by Neighborhood

In [132]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Aquarium  Art Gallery  \
0                 0                    0         0            0   
1                 0                    0         0            0   
2                 0                    0         0            0   
3                 0                    0         0            0   
4                 0                    0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Basketball Court  Basketball Stadium  Beer Bar  Beer Store  \
0               0                 0                   0         0           0   
1               0                 0                   0         0           0   
2               0                 0                   0         0           0   
3               0                 0                   0         0           0   
4               0                 0                   0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Burger Joint  \
0               0        0            0                0             0   
1               0        0            0                0             0   
2               0        0            0                0             0   
3               0        0            0                0             0   
4               0        0            0                0             0   

   Burrito Place  Bus Line  Bus Station  Business Service  Butcher  Café  \
0              0         0            0                 0        0     0   
1              0         0            0                 0        0     0   
2              0         0            0                 0        0     0   
3              0         0            0                 0        0     0   
4              0         0            0                 0        0     0   

   Cajun /

In [133]:
toronto_onehot.shape

(1273, 227)

#### Find the mean for each category by Neighborhood

In [134]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
9                                 Caledonia-Fairbanks     0.000000   
10                                          Cedarbrae     0.000000   
11                                 Central Bay Street     0.033333   
12                                           Christie     0.000000   
13                               Church and Wellesley     0.033333   
14  Clairville, Humberwood, Woodbine Downs, West H...     0.000000   
15            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
16    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
17                     Commerce Court, Victoria Hotel     0.000000   
18                                         Davisville     0.000000   
19                                   Davisville North     0.000000   
20   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
21                                    Don Mills North     0.000000   
22                                    Don Mills South     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                  Downsview Central     0.000000   
25                                     Downsview East     0.000000   
26                                Downsview Northwest     0.000000   
27                                     Downsview West     0.000000   
28                       Dufferin, Dovercourt Village     0.000000   
29                                     Enclave of L4W     0.000000   
30                                     Enclave of M4L     0.052632   
31                                     Enclave of M5E     0.000000   
32  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
33                       Fairview, Henry Farm, Oriole     0.000000   
34             First Canadian Place, Underground city     0.000000   
35                           Forest Hill North & West     0.000000   
36                           Garden District, Ryerson     0.000000   
37                                          Glencairn     0.000000   
38                    Golden Mile, Clairlea, Oakridge     0.000000   
39                  Guildwood, Morningside, West Hill     0.000000   
40  Harbourfront East, Union Station, Toronto Islands     0.000000   
41                      High Park, The Junction South     0.000000   
42                                  Hillcrest Village     0.000000   
43                                      Humber Summit     0.000000   
44                                   Humberlea, Emery     0.000000   
45                                 Humewood-Cedarvale     0.000000   
46                     India Bazaar, The Beaches West     0.000000   
47        Kennedy Park, Ionview, East Birchmount Park     0.000000   
48          Kensington Market, Chinatown, Grange Park     0.000000   
49  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
50                   Lawrence Manor, Lawrence Heights     0.000000   
51                                      Lawrence Park     0.000000   
52                                            Leaside     0.000000   
53                           Little Portugal, Trinity     0.033333   
54                                     Malvern, Rouge     0.000000   
55  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
56  Mimico NW, The Queensway West, South of Bloor,.

In [135]:
toronto_grouped.shape

(100, 227)

In [136]:
#### We look for the top 5 venues by Neighborhood'

In [137]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot   0.2
1                     Lounge   0.2
2             Clothing Store   0.2
3  Latin American Restaurant   0.2
4               Skating Rink   0.2


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.25
1     Coffee Shop  0.12
2    Skating Rink  0.12
3  Sandwich Place  0.12
4    Dance Studio  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
            venue  freq
0     Coffee Shop  0.13
1            Bank  0.13
2  Ice Cream Shop  0.07
3           Diner  0.07
4     Gas Station  0.07


----Bayview Village----
                     venue  freq
0                     Café  0.25
1                     Bank  0.25
2       Chinese Restaurant  0.25
3      Japanese Restaurant  0.25
4  New American Restaurant  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0  Italian Restaurant  0.09
1      Sandwich Place  0.09
2          Restaurant  0.09
3         Cof

#### Created a funtion to find the most common category of venue by Neighborhood

In [138]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [139]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                Lounge        Clothing Store  Latin American Restaurant   
1           Coffee Shop          Skating Rink             Sandwich Place   
2                  Bank        Ice Cream Shop                      Diner   
3                  Bank    Chinese Restaurant        Japanese Restaurant   
4        Sandwich Place            Restaurant                Coffee Shop   

     5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0             Skating Rink                Yoga Studio                Museum   
1             Dance Studio                        Pub            Playground   
2              Gas Station           Sushi Restaurant         Shopping Mall   
3  New American Restaurant  Middle Eastern Restaurant    Miscellaneous Shop   
4              Pizza Place       Fast Food Restaurant              Pharmacy   

       8th Most Common Venue       9th Most Common Venue  \
0  Middle Eastern Restaurant          Miscellaneous Shop   
1          Other Repair Shop               Movie Theater   
2             Sandwich Place                    Pharmacy   
3          Mobile Phone Shop  Modern European Restaurant   
4                    Butcher                        Café   

            10th Most Common Venue  
0                Mobile Phone Shop  
1               Mexican Restaurant  
2                             Park  
3  Molecular Gastronomy Restaurant  
4                     Liquor Store

#### Apply K-means clustering to the dataframe with 4 clusters

In [140]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1])

In [141]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = merged

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood',how ='inner')

toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels  1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656               1   Fast Food Restaurant                    Park   
1 -79.315572               0  Portuguese Restaurant           Grocery Store   
2 -79.360636               0            Coffee Shop                    Park   
3 -79.464763               0         Clothing Store  Furniture / Home Store   
4 -79.389494               0            Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
0     Food & Drink Shop                Museum  Mediterranean Restaurant   
1           Coffee Shop          Hockey Arena                Nail Salon   
2                Bakery                   Pub                Restaurant   
3     Accessories Store              Boutique        Miscellaneous Shop   
4         Burrito Place           Yoga Studio               Salad Place   

             6th Most Common Venue 7th Most Common Venue  \
0                    Metro Station    Mexican Restaurant   
1  Molecular Gastronomy Restaurant                Museum   
2                      Event Space        Chocolate Shop   
3            Vietnamese Restaurant           Coffee Shop   
4                     Burger Joint                  Café   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Middle Eastern Restaurant    Miscellaneous Shop      Mobile Phone Shop  
1              Movie Theater                 Motel    Monument / Landmark  
2          French Restaurant                  Bank                   Café  
3        Monument / Landmark           Music Venue                 Museum  
4                Music Venue        Sandwich Place             Restaurant

#### Plot a map with the 4 clusters

In [142]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Then review each cluster

In [143]:
_1= toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

cluster_1 = venues_count.iloc[:,0:1].join(_1.set_index('Neighborhood'), on='Neighborhood',how ='inner')
cluster_1.rename(columns={'Neighborhood Latitude': 'No of Venues'}, inplace=True)
cluster_1

No of Venues  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               15   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             23   
Berczy Park                                                   30   
Birch Cliff, Cliffside West                                    4   
Brockton, Parkdale Village, Exhibition Place                  22   
CN Tower, King and Spadina, Railway Lands, Harb...            17   
Cedarbrae                                                      7   
Central Bay Street                                            30   
Christie                                                      15   
Church and Wellesley                                          30   
Clairville, Humberwood, Woodbine Downs, West Hu...             4   
Clarks Corners, Tam O'Shanter, Sullivan                       13   
Cliffside, Cliffcrest, Scarborough Village West                2   
Commerce Court, Victoria Hotel                                30   
Davisville                                                    28   
Davisville North                                               9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn               4   
Don Mills North                                                5   
Don Mills South                                               16   
Dorset Park, Wexford Heights, Scarborough Town ...             6   
Downsview Central                                              3   
Downsview Northwest                                            6   
Dufferin, Dovercourt Village                                  14   
Enclave of L4W                                                14   
Enclave of M4L                                                19   
Enclave of M5E                                                30   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...             7   
Fairview, Henry Farm, Oriole                                  30   
First Canadian Place, Underground city                        30   
Forest Hill North & West                                       4   
Garden District, Ryerson                                      30   
Glencairn                                                      4   
Golden Mile, Clairlea, Oakridge                               10   
Guildwood, Morningside, West Hill                              9   
Harbourfront East, Union Station, Toronto Islands             30   
High Park, The Junction South                                 18   
Hillcrest Village                                              6   
Humber Summit                                                  2   
Humberlea, Emery                                               2   
Humewood-Cedarvale                                             4   
India Bazaar, The Beaches West                                19   
Kennedy Park, Ionview, East Birchmount Park                    7   
Kensington Market, Chinatown, Grange Park                     30   
Lawrence Manor, Lawrence Heights                               9   
Leaside                                                       26   
Little Portugal, Trinity                                      30   
Malvern, Rouge                                                 1   
Mimico NW, The Queensway West, South of Bloor, ...            12   
New Toronto, Mimico South, Humber Bay Shores                  11   
North Park, Maple Leaf Park, Upwood Park                       5   
North Toronto West                                            14   
Northwood Park, York University                                6   
Old Mill South, King's Mill Park, Sunnylea, Hum...             1   
Ontario Provincial Government                                 30   
Parkdale, Roncesvalles                  

In [144]:
_2= toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

cluster_2 = venues_count.iloc[:,0:1].join(_2.set_index('Neighborhood'), on='Neighborhood',how ='inner')
cluster_2.rename(columns={'Neighborhood Latitude': 'No of Venues'}, inplace=True)
cluster_2

No of Venues  \
Neighborhood                                                       
Caledonia-Fairbanks                                            4   
Downsview East                                                 4   
Downsview West                                                 4   
Kingsview Village, St. Phillips, Martin Grove G...             3   
Lawrence Park                                                  4   
Milliken, Agincourt North, Steeles East, L'Amor...             3   
Moore Park, Summerhill East                                    4   
Parkwoods                                                      3   
Rosedale                                                       4   
The Danforth  East                                             2   
Weston                                                         1   
Willowdale, Newtonbrook                                        1   
York Mills West                                                2   

                                                    Cluster Labels  \
Neighborhood                                                         
Caledonia-Fairbanks                                              1   
Downsview East                                                   1   
Downsview West                                                   1   
Kingsview Village, St. Phillips, Martin Grove G...               1   
Lawrence Park                                                    1   
Milliken, Agincourt North, Steeles East, L'Amor...               1   
Moore Park, Summerhill East                                      1   
Parkwoods                                                        1   
Rosedale                                                         1   
The Danforth  East                                               1   
Weston                                                           1   
Willowdale, Newtonbrook                                          1   
York Mills West                                                  1   

                                                   1st Most Common Venue  \
Neighborhood                                                               
Caledonia-Fairbanks                                                 Park   
Downsview East                                                   Airport   
Downsview West                                             Grocery Store   
Kingsview Village, St. Phillips, Martin Grove G...        Sandwich Place   
Lawrence Park                                                       Park   
Milliken, Agincourt North, Steeles East, L'Amor...            Playground   
Moore Park, Summerhill East                                   Playground   
Parkwoods                                           Fast Food Restaurant   
Rosedale                                                            Park   
The Danforth  East                                     Convenience Store   
Weston                                                              Park   
Willowdale, Newtonbrook                                             Park   
York Mills West                                        Convenience Store   

                                                   2nd Most Common Venue  \
Neighborhood                                                               
Caledonia-Fairbanks                                        Women's Store   
Downsview East                                                      Park   
Downsview West                                                      Park   
Kingsview Village, St. Phillips, Martin Grove G...                  Park   
Lawrence Park                                                Swim School   
Milliken, Agincourt North, Steeles East, L'Amor...                  Park   
Moore Park, Summerhill East                                         Park   
Parkwoods                                                           Park   
Rosedale                                                      Playground   
The

In [145]:
_3= toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

cluster_3 = venues_count.iloc[:,0:1].join(_3.set_index('Neighborhood'), on='Neighborhood',how ='inner')
cluster_3.rename(columns={'Neighborhood Latitude': 'No of Venues'}, inplace=True)
cluster_3

No of Venues  Cluster Labels  \
Neighborhood                                                           
Rouge Hill, Port Union, Highland Creek             1               2   

                                       1st Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek                   Bar   

                                       2nd Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek           Yoga Studio   

                                       3rd Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek           Music Venue   

                                       4th Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek         Metro Station   

                                       5th Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek    Mexican Restaurant   

                                            6th Most Common Venue  \
Neighborhood                                                        
Rouge Hill, Port Union, Highland Creek  Middle Eastern Restaurant   

                                       7th Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek    Miscellaneous Shop   

                                       8th Most Common Venue  \
Neighborhood                                                   
Rouge Hill, Port Union, Highland Creek     Mobile Phone Shop   

                                             9th Most Common Venue  \
Neighborhood                                                         
Rouge Hill, Port Union, Highland Creek  Modern European Restaurant   

                                                 10th Most Common Venue  
Neighborhood                                                             
Rouge Hill, Port Union, Highland Creek  Molecular Gastronomy Restaurant

In [146]:
_4= toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

cluster_4 = venues_count.iloc[:,0:1].join(_4.set_index('Neighborhood'), on='Neighborhood',how ='inner')
cluster_4.rename(columns={'Neighborhood Latitude': 'No of Venues'}, inplace=True)
cluster_4

No of Venues  \
Neighborhood                                                       
West Deane Park, Princess Gardens, Martin Grove...             1   

                                                    Cluster Labels  \
Neighborhood                                                         
West Deane Park, Princess Gardens, Martin Grove...               3   

                                                        1st Most Common Venue  \
Neighborhood                                                                    
West Deane Park, Princess Gardens, Martin Grove...  Middle Eastern Restaurant   

                                                   2nd Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...           Yoga Studio   

                                                   3rd Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...                 Plane   

                                                       4th Most Common Venue  \
Neighborhood                                                                   
West Deane Park, Princess Gardens, Martin Grove...  Mediterranean Restaurant   

                                                   5th Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...         Metro Station   

                                                   6th Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...    Mexican Restaurant   

                                                   7th Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...    Miscellaneous Shop   

                                                   8th Most Common Venue  \
Neighborhood                                                               
West Deane Park, Princess Gardens, Martin Grove...     Mobile Phone Shop   

                                                         9th Most Common Venue  \
Neighborhood                                                                     
West Deane Park, Princess Gardens, Martin Grove...  Modern European Restaurant   

                                                             10th Most Common Venue  
Neighborhood                                                                         
West Deane Park, Princess Gardens, Martin Grove...  Molecular Gastronomy Restaurant

### We can see the result is skewed because there are some Neighborhoods with just 1 close venue, 